onnxscript/ir/traversal.py

```diff
@@ -0,0 +1,85 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+"""Utilities for traversing the IR graph."""
+
+from __future__ import annotations
+
+__all__ = [
+    "RecursiveGraphIterator",
+]
+
+from typing import Callable, Iterator, Reversible
+
+from typing_extensions import Self
+
+from onnxscript.ir import _core, _enums
+
+
+class RecursiveGraphIterator(Iterator[_core.Node], Reversible[_core.Node]):
+    def __init__(
+        self,
+        graph: _core.Graph | _core.Function | _core.GraphView,
+        *,
+        recursive: Callable[[_core.Node], bool] | None = None,
+        reverse: bool = False,
+    ):
+        """Iterate over the nodes in the graph, recursively visiting subgraphs.
+
+        Args:
+            graph: The graph to traverse.
+            recursive: A callback that determines whether to recursively visit the subgraphs
+                contained in a node. If not provided, all nodes in subgraphs are visited.
+            reverse: Whether to iterate in reverse order.
+        """
+        self._graph = graph
+        self._recursive = recursive
+        self._reverse = reverse
+        self._iterator = self._recursive_node_iter(graph)
+
+    def __iter__(self) -> Self:
+        self._iterator = self._recursive_node_iter(self._graph)
+        return self
+
+    def __next__(self) -> _core.Node:
+        return next(self._iterator)
+
+    def _recursive_node_iter(
+        self, graph: _core.Graph | _core.Function | _core.GraphView
+    ) -> Iterator[_core.Node]:
+        iterable = reversed(graph) if self._reverse else graph
+        for node in iterable:  # type: ignore[union-attr]
+            yield node
+            if self._recursive is not None and not self._recursive(node):
+                continue
+            yield from self._iterate_subgraphs(node)
+
+    def _iterate_subgraphs(self, node: _core.Node):
+        iterator = (
+            reversed(node.attributes.values()) if self._reverse else node.attributes.values()
```

Comment:
This may be unnecessary. This will change things only for a IF-node. When users want to iterate forward or backward, I don't think they care what order the THEN-branch and ELSE-branch are visited in (as long as internally those branches are visited in the right order). Semantically, it is a reasonable spec that the two branches may be visited in either order.

If there is a cost to reversing the attributes, we may as well avoid it.

Reply:
What users may care more about is pre-order or post-order: are the THEN/ELSE branches visited before or after the IF node itself. But that would mean another option ... so, perhaps better to wait for a use-case before bothering about it.

Reply:
Done